pkg/server/commands/reverseexpand/reverse_expand_weighted.go

```diff
@@ -491,3 +529,232 @@ func (c *ReverseExpandQuery) buildFilteredIterator(
+	edges []*weightedGraph.WeightedAuthorizationModelEdge,
+	sourceUserType string,
+	resolutionMetadata *ResolutionMetadata,
+) {
+	// We need to create a new stack with the top item from the original request's stack
+	// and use it to get the candidates for the lowest weight edge.
+	// If the edge is a tuple to userset edge, we need to later check the candidates against the
+	// original relationStack with the top item removed.
+	var topItemStack stack.Stack[typeRelEntry]
+	if req.relationStack != nil {
+		topItem, newStack := stack.Pop(req.relationStack)
+		req.relationStack = newStack
+		topItemStack = stack.Push(nil, topItem)
+	}
+
+	// getting list object candidates from the lowest weight edge and have its result
+	// pass through tmpResultChan.
+	pool.Go(func(ctx context.Context) error {
+		defer close(tmpResultChan)
+		// stack with only the top item in it
+		newReq := req.clone()
+		newReq.relationStack = topItemStack
+		err := c.shallowClone().loopOverEdges(
+			ctx,
+			newReq,
+			edges,
+			false,
+			resolutionMetadata,
+			tmpResultChan,
+			sourceUserType,
+		)
+		return err
+	})
+}
+
+// callCheckForCandidates calls check on the list objects candidates against non lowest weight edges.
+func (c *ReverseExpandQuery) callCheckForCandidates(
+	pool *concurrency.Pool,
+	req *ReverseExpandRequest,
+	tmpResultChan <-chan *ReverseExpandResult,
+	resultChan chan<- *ReverseExpandResult,
+	userset *openfgav1.Userset,
+	isAllowed bool,
+) {
+	pool.Go(func(ctx context.Context) error {
+		for tmpResult := range tmpResultChan {
+			handlerFunc := c.localCheckResolver.CheckRewrite(ctx,
+				&graph.ResolveCheckRequest{
+					StoreID:              req.StoreID,
+					AuthorizationModelID: c.typesystem.GetAuthorizationModelID(),
+					TupleKey:             tuple.NewTupleKey(tmpResult.Object, req.Relation, req.User.String()),
+					ContextualTuples:     req.ContextualTuples,
+					Context:              req.Context,
+					Consistency:          req.Consistency,
+					RequestMetadata:      graph.NewCheckRequestMetadata(),
+				}, userset)
+			tmpCheckResult, err := handlerFunc(ctx)
+			if err != nil {
+				functionName := "intersectionHandler"
+				if !isAllowed {
+					functionName = "exclusionHandler"
+				}
+				c.logger.Error("Failed to execute", zap.Error(err),
+					zap.String("function", functionName),
+					zap.String("object", tmpResult.Object),
+					zap.String("relation", req.Relation),
+					zap.String("user", req.User.String()))
+				return err
+			}
+
+			// Check if the check result matches the expect isAllowed value
+			// eg, for intersection we expect the check result to be true
+			// and for exclusion we expect the check result to be false.
+			if tmpCheckResult.GetAllowed() == isAllowed {
+				if stack.Len(req.relationStack) == 0 {
+					// If the original stack only had 1 value, we can trySendCandidate right away (nothing more to check)
+					c.trySendCandidate(ctx, false, tmpResult.Object, resultChan)
+				} else {
```

Comment:
Is there an opportunity to remove this else here as well and simplify the complexity by continuing within the if above?